### PR TITLE
[google_maps] Prepares packages to endorse web.

### PR DESCRIPTION
**Pull request by ditman**
Prepares the google_maps packages to endorse the Web platform:

* `[google_maps_flutter]` Changes `example/integration_test` to run in the web.
  * Splits the only file there in 3 for them to be slightly more manageable.
  * (Does not require publishing)
* `[google_maps_flutter_platform_interface]` Adds a test coming from the core plugin to the unit tests of this package.
  * (Does not require publishing)
* `[google_maps_flutter_web]` Changes to add an "inspector" object, and to conform with the tests in the core package.
  * Implements a `GoogleMapsInspectorPlatform` to allow integration tests to inspect parts of the internal state of a gmap.
  * Fires a `MapStyleException` when an invalid JSON is used in `setMapStyle` (was `FormatException` previously), to conform with the expected behavior in the core plugin tests.
  * (Requires publishing)

## Issues

* Part of: https://github.com/flutter/flutter/issues/80688

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by ditman** [comment]
I thought of renaming the files midway through, and the commits ended up being a little bit sloppy.

This is what an average "fix" of the core test looks like: https://github.com/flutter/packages/pull/4064/commits/31d51ca3ac65e1f6eb345a1116bfc97620871e57

Maps on the web need a little bit more scaffolding to render than in mobile, apparently.

**Comment by ditman** [comment]
```
Testing example/integration_test/tiles_inspector_test.dart...
Running command: "/tmp/cirrus-ci-build/packages/google_maps_flutter/google_maps_flutter/example/android/gradlew app:assembleDebug -Pverbose=true -Ptarget=/tmp/cirrus-ci-build/packages/google_maps_flutter/google_maps_flutter/example/integration_test/tiles_inspector_test.dart" in /tmp/cirrus-ci-build/packages/google_maps_flutter/google_maps_flutter/example/android
```

Does this mean that for each file that we have in `integration_test`, we create a new FTL run? :S

**Comment by ditman** [comment]
All that would be needed to endorse the web package after landing this PR is something like this:

https://github.com/flutter/packages/compare/23f2c5e...ditman:flutter-packages:maps_endorse_web

**Comment by stuartmorgan** [comment]
> All that would be needed to endorse the web package after landing this PR is something like this:
> 
> [23f2c5e...ditman:flutter-packages:maps_endorse_web](https://github.com/flutter/packages/compare/23f2c5e...ditman:flutter-packages:maps_endorse_web)

We'll want to remember to update the web package's README to say it's endorsed when we do that. IIRC the tooling will let us do it in the same PR since it's not changing Dart code.

File: packages/google_maps_flutter/google_maps_flutter_web/CHANGELOG.md
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter_web/CHANGELOG.md:4]: Shouldn't we mark this as a breaking change and version accordingly?

File: packages/google_maps_flutter/google_maps_flutter_web/lib/src/google_maps_flutter_web.dart
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter_web/lib/src/google_maps_flutter_web.dart:336]: I would suggest keeping `_map` private, and passing a lookup function here to replace the `GoogleMapController _map(int mapId) =>
 (GoogleMapsFlutterPlatform.instance as GoogleMapsPlugin).map(mapId)` in the inspector.

See
https://github.com/flutter/packages/blob/fba97fa4cc86a2100a06a58205a5ccb366101cdc/packages/google_maps_flutter/google_maps_flutter_ios/lib/src/google_map_inspector_ios.dart#L12-L15
and
https://github.com/flutter/packages/blob/fba97fa4cc86a2100a06a58205a5ccb366101cdc/packages/google_maps_flutter/google_maps_flutter_ios/lib/src/google_maps_flutter_ios.dart#L589-L592

File: packages/google_maps_flutter/google_maps_flutter_web/lib/src/google_maps_inspector_web.dart
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter_web/lib/src/google_maps_inspector_web.dart:5]: Why do we need this? That seems like a design concern.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]

File: packages/google_maps_flutter/google_maps_flutter/example/integration_test/tiles_inspector_test.dart
Comment (ditman) [reply packages/google_maps_flutter/google_maps_flutter/example/integration_test/tiles_inspector_test.dart:1]: The only changes there were added to make the map render in the web. The scaffolding around the map widgets in this test wasn't enough for the web implementation of the map to actually render.

File: packages/google_maps_flutter/google_maps_flutter_web/CHANGELOG.md
Comment (ditman) [reply packages/google_maps_flutter/google_maps_flutter_web/CHANGELOG.md:4]: We could, however this `_web` package isn't used directly by end users, right? I actually needed to do this change for a test in the core package to pass :)

File: packages/google_maps_flutter/google_maps_flutter_web/lib/src/google_maps_flutter_web.dart
Comment (ditman) [reply packages/google_maps_flutter/google_maps_flutter_web/lib/src/google_maps_flutter_web.dart:336]: Ah yes, I like this better, good call!

File: packages/google_maps_flutter/google_maps_flutter_web/lib/src/google_maps_inspector_web.dart
Comment (ditman) [reply packages/google_maps_flutter/google_maps_flutter_web/lib/src/google_maps_inspector_web.dart:5]: I defined some fields as `@visibleForTesting`, and the analyzer doesn't understand that this file is only used for testing (probably because it MAY not be used only for testing). The suggestion you gave above might allow me to remove this, however.

File: packages/google_maps_flutter/google_maps_flutter_web/CHANGELOG.md
Comment (ditman) [reply packages/google_maps_flutter/google_maps_flutter_web/CHANGELOG.md:4]: (Done)

File: packages/google_maps_flutter/google_maps_flutter_web/lib/src/google_maps_flutter_web.dart
Comment (ditman) [reply packages/google_maps_flutter/google_maps_flutter_web/lib/src/google_maps_flutter_web.dart:336]: (Done)

File: packages/google_maps_flutter/google_maps_flutter_web/lib/src/google_maps_inspector_web.dart
Comment (ditman) [reply packages/google_maps_flutter/google_maps_flutter_web/lib/src/google_maps_inspector_web.dart:5]: (Done)

**Comment by ditman** [comment]
Thanks for the review @stuartmorgan!

* I've used the "ConfigurationProvider" trick you suggested so I don't have to punch so many holes into the currently existing objects. This has greatly simplified the actual implementation bits!
* I've also versioned the `_web` change as **major** because of the change in the Exception thrown by the `setConfiguration` method.

> I'm assuming there are no changes as part of the move, so these "new" files don't need review?

Actually, I did some changes so the tests could run on the web. Highlights:

* Stop using `dart:io Platform` and use `defaultTargetPlatform`, [like so](https://github.com/flutter/packages/pull/4064/files#diff-351bac8f83723887bb1a67f4d5a0022a9cb61864ae558745187dcb8a8b46c470R18-R21).
* Change the scaffolding of the Map widget for it to render on the web, [like in this commit](https://github.com/flutter/packages/pull/4064/commits/31d51ca3ac65e1f6eb345a1116bfc97620871e57).
  * (The minimal code currently in the integration tests doesn't work on the web)
* Used `isWeb` to opt-in to some of the tests [like here](https://github.com/flutter/packages/pull/4064/files#diff-351bac8f83723887bb1a67f4d5a0022a9cb61864ae558745187dcb8a8b46c470R80), or opt-out when a feature is not supported, [like here](https://github.com/flutter/packages/pull/4064/files#diff-f5967ea8f677bdea6eaa62b49de86d211b434e5889ca127a0f34536f90231371R210).

**No `expect` calls/checks were tweaked in this PR, however.**

I also moved [this test](https://github.com/flutter/packages/pull/4064/files#diff-a8066076d538f1e1e21b27bac466302cb9317b643500ce81b991102a94611070R171) from the core plugin to the platform interface package (it didn't make much sense in the core plugin.)

**Comment by ditman** [comment]
QQ @stuartmorgan. Each `integration_test/*_test.dart` file in the core package **triggers an FTL run** when running for android. Given the flakiness of FTL, should I un-split the test files so there's only one FTL run when testing this package?

**Comment by stuartmorgan** [comment]
> QQ @stuartmorgan. Each `integration_test/*_test.dart` file in the core package **triggers an FTL run** when running for android. Given how FTL some times flakes on our CI, should I un-split the test files so there's only one FTL run when testing this package to minimize the number of consecutive FTL runs?

Oh right, I forgot about that. I'm not worried about increasing flake since the tooling currently reruns a failing FTL run, so this would potentially help in that regard, but it does look like this just makes the test way slower overall, so we should recombine them for now. Maybe we can revisit once we have emulator tests, where I don't think we'll have to do that.

File: packages/google_maps_flutter/google_maps_flutter_web/lib/src/google_maps_controller.dart
Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter_web/lib/src/google_maps_controller.dart:61]: Maybe say "for the inspector" rather than "for integration tests" so it's clearer why it's not annotated as test-only?

Comment (stuartmorgan) [inline packages/google_maps_flutter/google_maps_flutter_web/lib/src/google_maps_controller.dart:61]: Also, nit: missing period.

**Review by stuartmorgan — APPROVED** [review]
LGTM modulo recombining the tests.

File: packages/google_maps_flutter/google_maps_flutter_web/lib/src/google_maps_controller.dart
Comment (ditman) [reply packages/google_maps_flutter/google_maps_flutter_web/lib/src/google_maps_controller.dart:61]: Should I push this? https://github.com/dart-lang/linter/pull/3820 :P

**Comment by ditman** [comment]
Will recombine the tests and land this after CI is happy. Thanks!

**Comment by ditman** [comment]
I think the repo_checks is tripping by the lack of CHANGELOG in the gogle_maps_flutter package, but all the changes happened in google_maps/example/integration_test.

Adding a `changelog: override` tag (FYI @stuartmorgan)

**Comment by stuartmorgan** [comment]
It's the run_test.sh script; that's not a standard thing so the tooling doesn't know to exclude it.

Having a generically named script in the main package that's actually web-only isn't ideal; I didn't notice that before. We should rename and comment-document it in the second PR.

**Comment by ditman** [comment]
> It's the run_test.sh script; that's not a standard thing so the tooling doesn't know to exclude it.

I ended up removing the bespoke scripts, here: https://github.com/flutter/packages/pull/4129 (now they live in my `~/bin`)

(If anybody ever other than me needs them, we can figure out a way to bring them back more properly).

**Pull request merged**
